[BUG] test_broadcast_nested_loop_join_special_case fails on databricks

Running integration tests on Databricks, I'm seeing test_broadcast_nested_loop_join_special_case fail:

FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[String][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Byte][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Short][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Integer][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Long][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Boolean][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Date][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Timestamp][IGNORE_ORDER({'local': True})]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Float][IGNORE_ORDER({'local': True}), INCOMPAT]
FAILED src/main/python/join_test.py::test_broadcast_nested_loop_join_special_case[Double][IGNORE_ORDER({'local': True}), INCOMPAT]

--- a/integration_tests/src/main/python/join_test.py
+++ b/integration_tests/src/main/python/join_test.py
@@ -97,8 +97,6 @@ def test_broadcast_nested_loop_join(data_gen):
 
 # local sort because of https://github.com/NVIDIA/spark-rapids/issues/84
 @ignore_order(local=True)
-@pytest.mark.xfail(condition=is_databricks_runtime(),
-    reason='https://github.com/NVIDIA/spark-rapids/issues/441')
 @pytest.mark.parametrize('data_gen', all_gen, ids=idfn)
 def test_broadcast_nested_loop_join_special_case(data_gen):
     def do_join(spark):

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuCartesianProductExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuCartesianProductExec.scala
@@ -18,7 +18,7 @@ package org.apache.spark.sql.rapids
 
 import java.io.{IOException, ObjectInputStream, ObjectOutputStream}
 
-import ai.rapids.cudf.{JCudfSerialization, NvtxColor, NvtxRange}
+import ai.rapids.cudf.{JCudfSerialization, NvtxColor, NvtxRange, Table}
 import com.nvidia.spark.rapids.{Arm, GpuBindReferences, GpuBuildLeft, GpuColumnarBatchSerializer, GpuColumnVector, GpuExec, GpuExpression, GpuSemaphore}
 import com.nvidia.spark.rapids.RapidsPluginImplicits._
 
@@ -177,8 +177,9 @@ class GpuCartesianRDD(
   }
 }
 
-object GpuNoColumnCrossJoin {
-  def divideIntoBatches(rowCounts: RDD[Long],
+object GpuNoColumnCrossJoin extends Arm {
+  def divideIntoBatches(
+      rowCounts: RDD[Long],
       targetSizeBytes: Long,
       numOutputRows: SQLMetric,
       numOutputBatches: SQLMetric): RDD[ColumnarBatch] = {
@@ -203,6 +204,22 @@ object GpuNoColumnCrossJoin {
     }
 
     rowCounts.flatMap(divideIntoBatches)
+  }
+
+  def divideIntoBatches(
+      table: Table,
+      numTimes: Long,
+      targetSizeBytes: Long,
+      numOutputRows: SQLMetric,
+      numOutputBatches: SQLMetric): Iterator[ColumnarBatch] = {
+    // TODO if we hit a point where we need to we can divide the data up into batches
+    //  The current use case is likely to be small enough that we are OK without this.
+    assert(numTimes < Int.MaxValue)
+    withResource(table.repeat(numTimes.toInt)) { repeated =>
+      numOutputBatches += 1
+      numOutputRows += repeated.getRowCount
+      Iterator(GpuColumnVector.from(repeated))
+    }
   }
 }
 

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala
@@ -21,6 +21,7 @@ import com.nvidia.spark.rapids._
 import com.nvidia.spark.rapids.GpuMetricNames.{NUM_OUTPUT_BATCHES, NUM_OUTPUT_ROWS, TOTAL_TIME}
 
 import org.apache.spark.TaskContext
+import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
@@ -191,6 +192,32 @@ abstract class GpuBroadcastNestedLoopJoinExecBase(
     }
   }
 
+  private[this] def makeBuiltTable(
+      broadcastRelation: Broadcast[SerializeConcatHostBuffersDeserializeBatch],
+      buildTime: SQLMetric,
+      buildDataSize: SQLMetric): Table = {
+    withResource(new NvtxWithMetrics("build join table", NvtxColor.GREEN, buildTime)) { _ =>
+      val ret = GpuColumnVector.from(broadcastRelation.value.batch)
+      // Don't warn for a leak, because we cannot control when we are done with this
+      (0 until ret.getNumberOfColumns).foreach(i => {
+        val column = ret.getColumn(i)
+        column.noWarnLeakExpected()
+        buildDataSize += column.getDeviceMemorySize
+      })
+      ret
+    }
+  }
+
+  private[this] def computeBuildRowCount(
+      broadcastRelation: Broadcast[SerializeConcatHostBuffersDeserializeBatch],
+      buildTime: SQLMetric,
+      buildDataSize: SQLMetric): Int = {
+    withResource(new NvtxWithMetrics("build join table", NvtxColor.GREEN, buildTime)) { _ =>
+      buildDataSize += 0
+      broadcastRelation.value.batch.numRows()
+    }
+  }
+
   override def doExecuteColumnar(): RDD[ColumnarBatch] = {
     val numOutputRows = longMetric(NUM_OUTPUT_ROWS)
     val numOutputBatches = longMetric(NUM_OUTPUT_BATCHES)
@@ -204,17 +231,19 @@ abstract class GpuBroadcastNestedLoopJoinExecBase(
     val buildTime = longMetric("buildTime")
     val buildDataSize = longMetric("buildDataSize")
 
-    val broadcastedRelation =
+    joinType match {
+      case _: InnerLike => // The only thing we support right now
+      case _ => throw new IllegalArgumentException(s"$joinType + $getGpuBuildSide is not" +
+          " supported and should be run on the CPU")
+    }
+
+    val broadcastRelation =
       broadcastExchange.executeColumnarBroadcast[SerializeConcatHostBuffersDeserializeBatch]()
 
     if (output.isEmpty) {
       assert(boundCondition.isEmpty)
 
-      lazy val buildCount: Long = {
-        withResource(new NvtxWithMetrics("build join table", NvtxColor.GREEN, buildTime)) { _ =>
-          broadcastedRelation.value.batch.numRows()
-        }
-      }
+      lazy val buildCount: Int = computeBuildRowCount(broadcastRelation, buildTime, buildDataSize)
 
       def getRowCountAndClose(cb: ColumnarBatch): Long = {
         val ret = cb.numRows()
@@ -229,28 +258,46 @@ abstract class GpuBroadcastNestedLoopJoinExecBase(
         targetSizeBytes,
         numOutputRows,
         numOutputBatches)
-    } else {
-      lazy val builtTable: Table = {
-        withResource(new NvtxWithMetrics("build join table", NvtxColor.GREEN, buildTime)) { _ =>
-          val ret = GpuColumnVector.from(broadcastedRelation.value.batch)
-          // Don't warn for a leak, because we cannot control when we are done with this
-          (0 until ret.getNumberOfColumns).foreach(i => {
-            val column = ret.getColumn(i)
-            column.noWarnLeakExpected()
-            buildDataSize += column.getDeviceMemorySize
-          })
-          ret
-        }
-      }
+    } else if (broadcast.output.isEmpty) {
+      assert(boundCondition.isEmpty)
+
+      lazy val buildCount: Int = computeBuildRowCount(broadcastRelation, buildTime, buildDataSize)
 
       streamed.executeColumnar().mapPartitions { streamedIter =>
-        joinType match {
-          case _: InnerLike => GpuBroadcastNestedLoopJoinExecBase.innerLikeJoin(streamedIter,
-            builtTable, getGpuBuildSide, boundCondition,
-            joinTime, joinOutputRows, numOutputRows, numOutputBatches, filterTime, totalTime)
-          case _ => throw new IllegalArgumentException(s"$joinType + $getGpuBuildSide is not" +
-            " supported and should be run on the CPU")
+        streamedIter.flatMap { cb =>
+          withResource(cb) { cb =>
+            withResource(GpuColumnVector.from(cb)) { table =>
+              GpuNoColumnCrossJoin.divideIntoBatches(
+                table,
+                buildCount,
+                targetSizeBytes,
+                numOutputRows,
+                numOutputBatches)
+            }
+          }
         }
+      }
+    } else if (streamed.output.isEmpty) {
+      assert(boundCondition.isEmpty)
+
+      // streamed is empty, not sure if this ever actually happens though
+      lazy val builtTable: Table = makeBuiltTable(broadcastRelation, buildTime, buildDataSize)
+      streamed.executeColumnar().flatMap { cb =>
+        withResource(cb) { cb =>
+          GpuNoColumnCrossJoin.divideIntoBatches(
+            builtTable,
+            cb.numRows(),
+            targetSizeBytes,
+            numOutputRows,
+            numOutputBatches)
+        }
+      }
+    } else {
+      lazy val builtTable: Table = makeBuiltTable(broadcastRelation, buildTime, buildDataSize)
+      streamed.executeColumnar().mapPartitions { streamedIter =>
+        GpuBroadcastNestedLoopJoinExecBase.innerLikeJoin(streamedIter,
+          builtTable, getGpuBuildSide, boundCondition,
+          joinTime, joinOutputRows, numOutputRows, numOutputBatches, filterTime, totalTime)
       }
     }
   }

